### PR TITLE
tele JB sync

### PR DESCRIPTION
**Pull request by valaises**

### CODE REVIEW

**Comment by olegklimov** [comment]
lsp.rs ->
    receive_workspace_changes.rs
    lsp.rs   -- calls receive_workspace_changes::xxx()
lsp_handlers.rs -> lsp_like_handlers.rs  -- calls receive_workspace_changes::xxx()

File: src/global_context.rs [hunk 1]
```diff
@@ -60,6 +66,7 @@ pub struct GlobalContext {
     pub telemetry: Arc<StdRwLock<telemetry_structs::Storage>>,
     pub vecdb_search: Arc<AMutex<Box<dyn VecdbSearch + Send>>>,
     pub ask_shutdown_sender: Arc<Mutex<std::sync::mpsc::Sender<String>>>,
+    pub lsp_backend_data: LSPBackendData,
```
Comment (reymondzzzz) [inline src/global_context.rs:69]: Name must be more representative. I see lsp_backend_data and don't understand for what it

File: src/http/routers/v1/lsp_like_handlers.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+pub async fn handle_v1_lsp_initialize(
+    Extension(global_context): Extension<SharedGlobalContext>,
+    body_bytes: hyper::body::Bytes,
+) -> Result<Response<Body>, ScratchError> {
+    let post = serde_json::from_slice::<PostInit>(&body_bytes).map_err(|e| {
+        ScratchError::new(StatusCode::BAD_REQUEST, format!("JSON problem: {}", e))
+    })?;
+
+    // Real work here
```
Comment (reymondzzzz) [inline src/http/routers/v1/lsp_like_handlers.rs:32]: delete comments

File: src/http/routers/v1/lsp_like_handlers.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        ScratchError::new(StatusCode::BAD_REQUEST, format!("JSON problem: {}", e))
+    })?;
+
+    receive_workspace_changes::on_did_change(
+        global_context.clone(),
+        &post.uri.to_string(),
+        &post.text,
+    ).await;
+
+    // Real work here
```
Comment (reymondzzzz) [inline src/http/routers/v1/lsp_like_handlers.rs:53]: here too

File: src/lsp.rs [hunk 4]
```diff
@@ -15,9 +13,11 @@ use tracing::{error, info};
 
 use crate::call_validation::{CodeCompletionInputs, CodeCompletionPost, CursorPosition, SamplingParameters};
 use crate::global_context;
-use crate::global_context::CommandLine;
+use crate::global_context::{CommandLine};
```
Comment (reymondzzzz) [inline src/lsp.rs:16]: ?

File: src/receive_workspace_changes.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+use tokio::sync::RwLock as ARwLock;
+use tracing::{info};
+
+use crate::global_context;
+use crate::telemetry;
+
+
+
+#[derive(Debug)]
+pub struct Document {
```
Comment (reymondzzzz) [inline src/receive_workspace_changes.rs:16]: why do u move Document here?

Comment (valaises) [reply src/receive_workspace_changes.rs:16]: bc Document is not being used anywhere else but in receive_workspace_changes

File: src/telemetry/basic_robot_human.rs [hunk 6]
```diff
@@ -27,12 +27,15 @@ pub fn increase_counters_from_finished_snippet(
     // Snippet is finished when it stops being valid for correction (user has changed code in a different place) or it timeouts
     fn robot_characters(snip: &SnippetTracker) -> i64 {
         let re = Regex::new(r"\s+").unwrap();
-        (re.replace_all(&snip.grey_text, "").len() as f64 * snip.remaining_percentage) as i64
+        let robot_characters = re.replace_all(&snip.grey_text, "").len() as i64;
+        info!("snippet: {}; robot_characters: {}", snip.grey_text, robot_characters);
+        robot_characters
```
Comment (olegklimov) [inline src/telemetry/basic_robot_human.rs:32]: If you need this log, introduce yourself: "increase_counters_from_finished_snippet: ........"  Then give more information what these numbers mean.

File: src/telemetry/utils.rs [hunk 7]
```diff
@@ -207,7 +208,11 @@ pub fn unchanged_percentage(
     text_a: &String,
     text_b: &String,
 ) -> f64 {
-    let diff = TextDiff::from_chars(text_a, text_b);
+    let re = Regex::new(r"\s+").unwrap();
+    let text_a = re.replace_all(text_a, "").to_string();
+    let text_b = re.replace_all(text_b, "").to_string();
+
+    let diff = TextDiff::from_chars(&text_a, &text_b);
```
Comment (olegklimov) [inline src/telemetry/utils.rs:215]: This deletes all space and newlines 🤔

File: src/main.rs [hunk 8]
```diff
@@ -72,7 +73,8 @@ async fn main() {
     }
 
     background_tasks.abort().await;
-    info!("saving telemetry without sending, so should be quick");
+    info!("starting: saving telemetry without sending, so should be quick");
     basic_transmit::telemetry_full_cycle(gcx.clone(), true).await;
+    info!("done: telemetry saving");
```
Comment (olegklimov) [inline src/main.rs:78]: can you undo this pls

**Review by olegklimov — APPROVED** [review]

**Pull request merged**